tree/ntuple/v7/src/RNTupleMerger.cxx

```diff
@@ -116,266 +133,572 @@ try {
+      if (colRangeCompressionSettings != 0)
+         sealedPageBuffers.resize(sealedPageBuffers.size() + pages.fPageInfos.size());
+
+      // Loop over the pages
+      std::uint64_t pageIdx = 0;
+      for (const auto &pageInfo : pages.fPageInfos) {
+         assert(pageIdx < sealedPages.size());
+         assert(sealedPageBuffers.size() == 0 || pageIdx < sealedPageBuffers.size());
+
+         ROnDiskPage::Key key{columnId, pageIdx};
+         auto onDiskPage = cluster->GetOnDiskPage(key);
+
+         const auto checksumSize = pageInfo.fHasChecksum * RPageStorage::kNBytesPageChecksum;
+         RPageStorage::RSealedPage &sealedPage = sealedPages[pageIdx];
+         sealedPage.SetNElements(pageInfo.fNElements);
+         sealedPage.SetHasChecksum(pageInfo.fHasChecksum);
+         sealedPage.SetBufferSize(pageInfo.fLocator.fBytesOnStorage + checksumSize);
+         sealedPage.SetBuffer(onDiskPage->GetAddress());
+         // TODO(gparolini): more graceful error handling (skip the page?)
+         sealedPage.VerifyChecksumIfEnabled().ThrowOnError();
+         R__ASSERT(onDiskPage && (onDiskPage->GetSize() == sealedPage.GetBufferSize()));
+
+         if (needsCompressionChange) {
+            RNTupleMerger::RChangeCompressionFunc compressTask{
+               pageIdx,          pageBufferBaseIdx, checksumSize,      colRangeCompressionSettings,
+               column.fOutputId, *srcColElement,    *dstColElement,    pageInfo,
+               options,          sealedPage,        sealedPageBuffers, source};
+
+            if (taskGroup)
+               taskGroup->Run(compressTask);
+            else
+               compressTask();
+         }
+
+         ++pageIdx;
+
+      } // end of loop over pages
+
+      if (taskGroup)
+         taskGroup->Wait();
+
+      sealedPagesV.push_back(std::move(sealedPages));
+      sealedPageGroups.emplace_back(column.fOutputId, sealedPagesV.back().cbegin(), sealedPagesV.back().cend());
+   } // end loop over common columns
+}
+
+// Generates default values for columns that are not present in the current source RNTuple
+// but are present in the destination's schema.
+static void GenerateExtraDstColumns(size_t nClusterEntries, std::span<RColumnInfo> extraDstColumns,
+                                    std::deque<RPageStorage::SealedPageSequence_t> &sealedPagesV,
+                                    std::vector<RPageStorage::RSealedPageGroup> &sealedPageGroups,
+                                    std::vector<std::unique_ptr<unsigned char[]>> &sealedPageBuffers,
+                                    const RNTupleDescriptor &srcDescriptor, const RNTupleDescriptor &dstDescriptor)
+{
+   for (const auto &column : extraDstColumns) {
+      const auto &columnId = column.fInputId;
+      const auto &columnDesc = dstDescriptor.GetColumnDescriptor(columnId);
+
+      // Check if this column is a child of a Collection or a Variant. If so, it has no data
+      // and can be skipped.
```

Comment:
We also have to skip auxiliary columns (non-principal columns of fields with multiple columns). I.e. the char column of `std::string`.